### ISSUE TITLE
Question: Proper use of save and restore layout

### ISSUE THREAD

**Issue by sopsick** [jupyterlab/lumino#79]
I have a fairly complex layout of widgets including some that are React widgets (i.e., wrapped in `UseSignal`), and multiple `datagrid` tabs with custom data models and custom selection models. When I leave the page I use `saveLayout` from the main `DockPanel`. When I come back to the page, I try to use `restoreLayout`. It does seem to restore the panels to the right positions, but the panels no longer work properly. 

I thought the idea was that when I am attempting to recreate the page, I build the widgets and then tell the `DockPanel` to `restoreLayout` and it would put them back in the last saved positions, but I am beginning to think I am not using this properly as it looks like the restore is attempting to recreate the widgets and mine are detached from the parent.

Any advice on how to restore the layout would be appreciated.

**Comment by afshin** [comment]
Hello, @sopsick!

I think one place to start is to take what `saveLayout` gives you, clear your page, then give that object back to `restoreLayout` to make sure it is working for you. What you describe should be possible and is indeed the purpose of those two methods.

For an example of a complex application that uses those methods to restore itself, check out this comment in JupyterLab that describes the layout restoration process. Keep in mind that it introduces some concepts you probably don't need (like a widget tracker), but the underlying work being done is restoring a complex set of widgets when a user refreshes the page.

cf. https://github.com/jupyterlab/jupyterlab/blob/master/packages/application/src/layoutrestorer.ts#L67-L115

**Comment by sopsick** [comment]
Just posting an update. I was able to get this working eventually. One of the complexities is the fact that some of my widgets are wrapped React Components (e.g., a Stacked Panel holding a  React Component) . As near as I can tell, React was not happy with the components being "alive" in the saved layout and not being part of the DOM. I had to dispose  of the React Widget when leaving the page without disposing its holding panel and then recreate the component on the reconstituted panel when I restored the layout. This allows the component to appear in the proper location in the DockPanel.
